crates/agent-sql/src/directives/beta_onboard.rs

```diff
@@ -41,39 +41,89 @@ pub async fn tenant_exists(
     Ok(exists.is_some())
 }
 
-/*
-TODO(johnny): A fast-follow will be inserting `ops/` specifications for the user:
-insert into live_specs (catalog_name, last_build_id, last_pub_id, spec, spec_type) values
-    ('ops/' || $2 || 'logs',  '00:00:00:00:00:00:00:00', '00:00:00:00:00:00:00:00', $4, 'collection'),
-    ('ops/' || $2 || 'stats', '00:00:00:00:00:00:00:00', '00:00:00:00:00:00:00:00', $5, 'collection')
-*/
-pub async fn provision_user(
-    user_id: Uuid,
-    tenant: &str,
+// ProvisionedTenant is the shape of a provisioned tenant.
+#[derive(Debug)]
+pub struct ProvisionedTenant {
+    // Draft into which provisioned catalog specs should be placed.
+    // It will be queued for publication upon the commit of this transaction.
+    pub draft_id: Id,
+}
+
+pub async fn provision_tenant(
+    accounts_user_email: &str,
     detail: Option<String>,
+    tenant: &str,
+    tenant_user_id: Uuid,
     txn: &mut sqlx::Transaction<'_, sqlx::Postgres>,
-) -> sqlx::Result<()> {
+) -> sqlx::Result<ProvisionedTenant> {
     let prefix = format!("{tenant}/");
 
-    sqlx::query!(
-        r#"with s1 as (
+    let provisioned = sqlx::query_as!(
+        ProvisionedTenant,
+        r#"with
+        accounts_root_user as (
+            -- Precondition: the accounts root user must exist.
+            -- Use a sub-select to select either one match or an explicit null row,
+            -- which will then fail a not-null constraint.
+            select (select id from auth.users where email = $4 limit 1) as accounts_id
+        ),
+        create_tenant as (
+            insert into tenants (tenant, detail) values ($2, $3)
+        ),
+        grant_user_admin_to_tenant as (
             insert into user_grants (user_id, object_role, capability, detail) values
                 ($1, $2, 'admin', $3)
         ),
-        s2 as (
+        grant_to_tenant as (
             insert into role_grants (subject_role, object_role, capability, detail) values
-                ($2, 'ops/' || $2, 'read', $3)
+                ($2, $2, 'write', $3),              -- Tenant specs may write to other tenant specs.
+                ($2, 'ops/' || $2, 'read', $3),     -- Tenant may read `ops/$tenant/...` collections.
+                ($2, 'estuary/public/', 'read', $3) -- Tenant may read `estuary/pubic/` collections.
+        ),
+        create_storage_mappings as (
+            insert into storage_mappings (catalog_prefix, spec, detail) values
+                ($2, '{"stores": [{"provider": "GCS", "bucket": "estuary-trial"}]}', $3),
```

Comment:
It feels weird to hard-code the storage bucket for this, especially given the fact that the accounts email is parameterized. Long term, I'd think that the storage bucket would be determined based on which data-plane the tenant will be mapped to. I don't know if it's reasonable to go straight to looking up the bucket based on the data-plane mapping, but it seems at least worth considering. If you did, it might also allow eliminate the storage mapping concerns that motivate doing this as an offline migration. Otherwise, perhaps just making it an application argument would be the way to go.

Reply:
:shrug: I'd like to KISS for now. Yes, this is a clear target for refactor when there's more built out to support creating storage mappings, etc.